// glob-sync/test/test.js
var path = require("path");
var glob = require("../").globSync;

// Test relative with leading ./
exports.relative = function(test){
  test.expect(1);

  var files = glob("*.js");
  
  test.deepEqual(files, [
    "index.js", 
    "test.js"
  ], "Correct array");

  test.done();
};

exports.relativeExpanded = function(test) {
  test.expect(1);

  var files = glob("foo/**/bar");
  
  test.deepEqual(files, [
    "foo/bar",
    "foo/baz/bar",
    "foo/baz/quux/bar"
  ], "Correct array");

  test.done();
};

exports.relativeExpandedFolder = function(test) {
  test.expect(1);

  var files = glob("foo/**/");

  test.deepEqual(files, [
    "foo/",
    "foo/baz/",
    "foo/baz/quux/"
  ], "Correct array");

  test.done();
};

exports.relativeFixed = function(test) {
  test.expect(1);

  var files = glob("foo/baz/quux/");

  test.deepEqual(files, [
    "foo/baz/quux/"
  ], "Correct array");

  test.done();
};

exports.relativeFullFixed = function(test) {
  test.expect(1);

  var files = glob("foo/baz/quux/bar");

  test.deepEqual(files, [
    "foo/baz/quux/bar"
  ], "Correct array");

  test.done();
};

exports.dotSlashRelativeFullFixed = function(test) {
  test.expect(1);

  var files = glob("./foo/baz/quux/bar");

  test.deepEqual(files, [
    "./foo/baz/quux/bar"
  ], "Correct array");

  test.done();
};

// Test absolute
exports.absolute = function(test){
  test.expect(1);

  var files = glob(path.resolve("*.js"));
  
  test.deepEqual(files, [
    path.resolve("index.js"), 
    path.resolve("test.js")
  ], "Correct array");

  test.done();
};
